site-permissions/site-permissions-impl/src/main/java/com/duckduckgo/site/permissions/impl/SitePermissionsRepository.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2022 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.site.permissions.impl
+
+import android.webkit.PermissionRequest
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.extractDomain
+import com.duckduckgo.di.scopes.ActivityScope
+import com.duckduckgo.site.permissions.store.sitepermissions.SitePermissionAskSettingType
+import com.duckduckgo.site.permissions.store.sitepermissions.SitePermissionsEntity
+import com.duckduckgo.site.permissions.store.SitePermissionsPreferencesImp
+import com.duckduckgo.site.permissions.store.sitepermissions.SitePermissionsDao
+import com.duckduckgo.site.permissions.store.sitepermissionsallowed.SitePermissionAllowedEntity
+import com.duckduckgo.site.permissions.store.sitepermissionsallowed.SitePermissionsAllowedDao
+import com.squareup.anvil.annotations.ContributesBinding
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+interface SitePermissionsRepository {
+    var askCameraEnabled: Boolean
+    var askMicEnabled: Boolean
+    fun isDomainAllowedToAsk(url: String, permission: String): Boolean
+    fun isDomainGranted(url: String, tabId: String, permission: String): Boolean
+    fun sitePermissionGranted(url: String, tabId: String, permission: String)
+    fun sitePermissionsWebsitesFlow(): Flow<List<SitePermissionsEntity>>
+    fun sitePermissionsForAllWebsites(): List<SitePermissionsEntity>
```

Comment:
NIT: should this be suspend and ensure it's executed in`io` thread or is that controlled externally? should we check we are not in main thread? what's the difference on threading between `sitePermissionsForAllWebsites`(non suspend) and `savePermission` (suspend)?

Reply:
My thought here was that suspend functions are called from a coroutine scope and the ones that are not, are directly called from a suspend fun. After talking to you I understood better the guidelines and I will follow them from now on 💯